nextest-runner/src/list/test_list.rs

```diff
@@ -882,6 +894,50 @@ pub(crate) fn make_test_command(
 
     let mut cmd = std::process::Command::new(program);
 
+    let make_windows_compatible_key = |key: OsString| {
+        // On Windows, environment variable keys are case-insensitive, so create
+        // a canonical form to check existence.
+        if cfg!(windows) {
+            key.to_ascii_lowercase()
+        } else {
+            key
+        }
+    };
+
+    // NB: we will always override user-provided environment variables with the
+    // ones we set below.
+    let mut existing_keys: HashSet<_> = std::env::vars_os()
+        .map(|(k, _v)| k)
+        .map(make_windows_compatible_key)
+        .collect();
+    for CargoEnvironmentVariable {
+        source,
+        name,
+        value,
+        force,
+        relative,
+    } in env
+    {
+        let name_os_string = make_windows_compatible_key(OsString::from(name));
+        if !existing_keys.insert(name_os_string) && !force {
+            continue;
+        }
```

Comment:
Is this the right order of things? What happens if there are multiple `[env]` vars defined in several config files, each with its own `force` setting?

My understanding is that the evaluation happens in two phases:

1. Configuration is all collected together, and lower-precedence configs are discarded.
2. Configuration is then applied.

The upshot of this is that if your cwd is `foo/bar`, and `foo/bar/.cargo/config.toml` is:

```toml
[env]
MY_VAR = { value = "foo-bar-config", force = true }
```

And `foo/.cargo/config.toml` is:

```toml
[env]
MY_VAR = { value = "foo-config", force = true }
```

Then Cargo will have `MY_VAR` set to `foo-bar-config`, whereas this code sets `MY_VAR` to `foo-config` if I've understood it correctly.

I'm not particularly sure what happens on Windows, though! It might be worth looking at Cargo's source code to get this right. I've had to do this several times for this code, and it's sometimes resulted in me filing bugs with upstream cargo (https://github.com/rust-lang/cargo/issues/10991, https://github.com/rust-lang/cargo/issues/10992).

Reply:
So... as far as I can tell this still isn't resolved. I added a bunch of assertions (search through my updates for `__NEXTEST_TESTING_EXTRA_CONFIG`) and they don't seem to be behaving as expected unfortunately.

I've addressed most of the other issues but I think this one is big enough that I'm going to leave it to you.

Reply:
@sunshowers I tried implementing this, but ran into some trouble. What exactly is the specification here? When passing extra configs via CLI, they need to override environment variables set anywhere else? Currently, when providing a file in this manner, it becomes indistinguishable from a config file which was discovered normally:

https://github.com/nextest-rs/nextest/blob/ffd00c9b92de878dfc1e5cca16d24f76d9e120d2/nextest-runner/src/cargo_config.rs#L302

So I would have to start tracking that information, yes?

In any event, does it really make sense for the extra configs to always override other things, even if `force` wasn't set to `true`? If I run `FOO=1 cargo nextest run --config path/to/config.toml`, I would personally expect `FOO` to take precedence even if there was a `FOO` set in `config.toml` without `force = true`.

Reply:
Hmm, I'm not sure what you mean here -- environment variables like `FOO=1` definitely should take precedence over config files that don't have `force = true` whether or not they're specified over the CLI. Config files over the CLI do take precedence over `.cargo/config` though (which should already be the case, you shouldn't have to touch it).

If you look at https://github.com/nextest-rs/nextest/pull/534/files#diff-b0286642fcb8773419c7282d5488e89e7fd9f2fe08c15c4cf5fe66c004f1bd24, all of the ones that don't have the environment variable set before execution (`__NEXTEST_TESTING_EXTRA_CONFIG_FORCE_IN_EXTRA` etc) should get set by `extra-config.toml`, and all of the ones that do have it set before execution (`__NEXTEST_TESTING_EXTRA_CONFIG_OVERRIDE_FORCE_IN_EXTRA` etc) should get set by `extra-config.toml` only if `force = true`. In all cases, `main-config.toml` should be completely ignored for these environment variables. (You can verify this yourself by setting the variables in the environment, then running `cargo test`.)

Does that make sense?

Reply:
> environment variables like FOO=1 definitely should take precedence over config files that don't have force = true whether or not they're specified over the CLI

Okay, I made an assumption about the environment variables based on the comment below, but it doesn't actually say what I thought it was, so this sounds makes sense.

I think I'm confused by this test case:

https://github.com/nextest-rs/nextest/blob/65977f9c8e948c5506e3e7f455b46458d9d76b31/fixtures/nextest-tests/.cargo/config#L28

According to my intuition, It seems like this should be `PASSED`, and we should take the value from the main config? (Actually, looking closer, neither the main config nor the extra config has a `PASSED` value...)

But according to the spec in this comment:

https://github.com/nextest-rs/nextest/blob/65977f9c8e948c5506e3e7f455b46458d9d76b31/fixtures/nextest-tests/.cargo/extra-config.toml#L4-L5

it should take on the value in the extra config, even though `force = true` in the main config.

Currently, it seems that there's no way to distinguish a cargo config that came from `config` versus `extra-config.toml`, so the only information we have to let us implement the specced behavior is the relative ordering of the config files. But, using the same rules as for environment variables and config files with different precedences discovered upwards in the directory tree, neither ordering would cause us to take on the value from the extra config file:

- If `config` is handled before `extra-config.toml`, then we set the value from the main config. When we see the value in `extra-config.toml` later, we don't use it, because we have already set a value.
- If `extra-config.toml` is handled before `config`, then we set the value from the extra config. But when we see the value in `config` later, we see that `force` is `true`, so we override it.

So in either case, the value from the main config takes precedence because `force` is `true`. I don't see an obvious way to keep the value from the extra config in this case. (Initially, I tried to check to see if the `source` is `CargoConfigSource::CliOption`, but this source is later lost when we read from a file rather than setting the config option directly on the command-line.)

It should be possible to preserve the information that a certain config value came from an extra config file (i.e. from a command-line option), but this isn't preserved currently IIUC, and implementing it seems to be a hassle, and I wonder if it would be the best behavior anyways.

Reply:
Can we just keep track of which variables we've seen and ignored in another hashset? Then we could just consult that and ignore any further attempts to set those variables in configs with later precedence.